types.go

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tally
+
+import "time"
+
+// Scope is a namespace wrapper around a stats reporter, ensuring that
+// all emitted values have a given prefix or set of tags
+type Scope interface {
+	// Counter returns the Counter object corresponding to the name
+	Counter(name string) Counter
+
+	// Gauge returns the Gauge object corresponding to the name
+	Gauge(name string) Gauge
+
+	// Timer returns the Timer object corresponding to the name
+	Timer(name string) Timer
+
+	// Tagged returns a new scope with the given tags
+	Tagged(tags map[string]string) Scope
+
+	// SubScope returns a new child scope with the given name
+	SubScope(name string) Scope
+
+	// Capabilities returns a description of metrics reporting capabilities
+	Capabilities() Capabilities
+
+	// Snapshot returns a copy of all values since the last report execution,
+	// this is an expensive operation and should only be use for testing purposes
+	Snapshot() Snapshot
+}
+
+// RootScope is a scope that manages itself and other Scopes
+type RootScope interface {
```

Comment:
I don't think this needs to be a public interface. Imaging another naive implementation of Scope that simply logs each metric call - it has no need for "close". We've had a similar proposal in OpenTracing and decided against having Close method in the interface, because in practice you need to call Close() around the same place where you call New(), meaning you actually have access to the real implementation, not the interface.

In jaeger tracer we had the tracer struct itself private, so we could not return it from the New() method, but we handled it by returning `(opentracing.Tracer, io.Closer)`, so that in the main function people do

```
func main() {
    tracer, closer := jaeger.New()
    defer closer.Close()
   . . .
}
```